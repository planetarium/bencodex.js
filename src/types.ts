/**
 * This module provides a basic type definitions for Bencodex, and some utility
 * functions for them.
 *
 * @module
 */

import { BencodexDictionary } from "./dict.ts";
import { areUint8ArraysEqual, compareUint8Arrays } from "./utils.ts";

/**
 * Represents a value which can be used as a key in a Bencodex dictionary.
 * It is either a string ("text" in Bencodex) or a `Unit8Array` instance
 * ("binary" in Bencodex).
 *
 * Note that every {@link Key} is also a {@link Value} (not vice versa).
 *
 * Here are examples of valid {@link Key} values:
 *
 * ```typescript
 * let text: Key = "key";
 * let binary: Key = new Uint8Array([0x6b, 0x65, 0x79]);
 * ```
 */
export type Key = string | Uint8Array;

/**
 * Represents a value which can be encoded in Bencodex.  Each data type in
 * Bencodex has its corresponding type in JavaScript:
 *
 * - `null` represents Bencodex's null
 * - `boolean` represents Bencodex's Boolean
 * - `bigint` represents Bencodex's integer (note that `number` is not a valid
 *   type for Bencodex's integer)
 * - `string` represents Bencodex's text
 * - `Uint8Array` represents Bencodex's binary
 * - `Value[]` represents Bencodex's list
 * - {@link Dictionary} represents Bencodex's dictionary (note that it is not
 *   a concrete type, but an interface)
 *
 * Here are examples of valid {@link Value}s:
 *
 * ```typescript
 * let null_: Value = null;
 * let bool: Value = true;
 * let integer: Value = 123n;  // note that it ends with suffix `n`
 * let text: Value = "value";
 * let binary: Value = new Uint8Array([0x76, 0x61, 0x6c, 0x75, 0x65]);
 * let list: Value = [
 *   null,
 *   true,
 *   123n,
 *   "value",
 *   new Uint8Array([0x76, 0x61, 0x6c, 0x75, 0x65]),
 *   [],
 *   new Map(),
 * ];
 * let dictionary: Value = new Map([
 *   ["string key", list],
 *   [new Uint8Array([0x6b, 0x65, 0x79]), "<- binary key"]
 * ]);
 * ```
 */
export type Value = null | boolean | bigint | Key | List | Dictionary;

/**
 * Represents a Bencodex list.  It basically is a read-only `Value[]`.
 */
export type List = readonly Value[];

/**
 * Represents a Bencodex dictionary.  It basically behaves like a read-only
 * `Map<Key, Value>`, but it is not necessarily a `Map` instance.
 *
 * @see {@link BencodexDictionary}
 * @see {@link RecordView}
 */
export interface Dictionary extends Iterable<readonly [Key, Value]> {
  /**
   * The number of key-value pairs in this dictionary.
   */
  readonly size: number;

  /**
   * Gets the value associated with the given key.
   * @param key The key to get the value associated with.
   * @returns The value associated with the given key, or `undefined` if the
   *         given key is not present in this dictionary.
   */
  get(key: Key): Value | undefined;

  /**
   * Checks if the given key is present in this dictionary.
   * @param key The key to check.
   * @returns `true` iff the given key is present in this dictionary.
   */
  has(key: Key): boolean;

  /**
   * Gets an iterable object which iterates over the keys in this dictionary.
   * @returns An iterable object which iterates over the keys in this
   *          dictionary.
   */
  keys(): Iterable<Key>;

  /**
   * Gets an iterable object which iterates over the values in this dictionary.
   * @returns An iterable object which iterates over the values in this
   *          dictionary.
   */
  values(): Iterable<Value>;

  /**
   * Gets an iterable object which iterates over the key-value pairs in this
   * dictionary.  This method is equivalent to `Dictionary[Symbol.iterator]()`.
   * @returns An iterable object which iterates over the key-value pairs in
   *          this dictionary.
   */
  entries(): Iterable<readonly [Key, Value]>;

  /**
   * Calls the given callback function for each key-value pair in this
   * dictionary.
   * @param callback The callback function to call for each key-value pair.
   *                 This function is called with three arguments: the value,
   *                 the key, and the dictionary itself.
   * @param thisArg The value to use as `this` when calling the callback
   *                function.
   */
  forEach(
    callback: (value: Value, key: Key, dictionary: Dictionary) => void,
    thisArg?: unknown,
  ): void;

  /**
   * Gets an iterator which iterates over the key-value pairs in this
   * dictionary.  This method is equivalent to `Dictionary.entries()`.
   * @returns An iterator which iterates over the key-value pairs in this
   *          dictionary.
   * @see {@link Dictionary.entries}
   */
  [Symbol.iterator](): Iterator<readonly [Key, Value]>;
}

/**
 * Checks if the given value is a Bencodex dictionary.
 *
 * In TypeScript, this function can be used as a type guard.
 * @param value The value to check.
 * @returns `true` iff the given value is a Bencodex dictionary.
 */
export function isDictionary(value: unknown): value is Dictionary {
  return (
    typeof value === "object" && value != null &&
    "size" in value && typeof value.size === "number" &&
    "get" in value && typeof value.get === "function" &&
    "has" in value && typeof value.has === "function" &&
    "keys" in value && typeof value.keys === "function" &&
    "values" in value && typeof value.values === "function" &&
    "entries" in value && typeof value.entries === "function" &&
    "forEach" in value && typeof value.forEach === "function" &&
    Symbol.iterator in value && typeof value[Symbol.iterator] === "function"
  );
}

/**
 * Checks if the given value can be used as a key in a Bencodex dictionary.
 *
 * In TypeScript, this function can be used as a type guard.
 * @param value The value to check.
 * @returns `true` iff the given value can be used as a key in a Bencodex
 *          dictionary.
 */
export function isKey(value: unknown): value is Key {
  return typeof value === "string" || value instanceof Uint8Array;
}

/**
 * Checks if the given keys have the same type and the same contents.  In other
 * words, this function checks if the given keys are encoded in the same
 * Bencodex data.
 * @param a A key to compare.
 * @param b Another key to compare.
 * @returns `true` iff the given keys have the same type and the same contents.
 *          It returns `false` when the given keys have invalid types even if
 *          they have the same contents, e.g., `areKeysEqual(1n, 1n)` returns
 *          `false`.
 */
export function areKeysEqual(a: Key, b: Key): boolean {
  if (typeof a === "string") {
    return typeof b === "string" && a === b;
  } else if (a instanceof Uint8Array) {
    return b instanceof Uint8Array && areUint8ArraysEqual(a, b);
  }
  return false;
}

/**
 * Compares two keys in the specified order in the Bencodex specification.
 *
 * This function can be passed to `Array.prototype.sort()` to sort keys in the
 * order specified in the Bencodex specification.
 * @param a A key to compare.
 * @param b Another key to compare.
 * @returns A negative number if `a` is former than `b`, zero if `a` is equal to
 *         `b`, or a positive number if `a` is latter than `b`.
 * @throws {TypeError} When any of the given keys is neither a `string` nor
 *         a `Uint8Array`.
 */
export function compareKeys(a: Key, b: Key): number {
  if (typeof a === "string") {
    if (typeof b === "string") return a < b ? -1 : a === b ? 0 : 1;
    else if (b instanceof Uint8Array) return 1;
    throw new TypeError(`Invalid key type: ${typeof b}`);
  } else if (a instanceof Uint8Array) {
    if (typeof b === "string") return -1;
    else if (b instanceof Uint8Array) return compareUint8Arrays(a, b);
    throw new TypeError(`Invalid key type: ${typeof b}`);
  }
  throw new TypeError(`Invalid key type: ${typeof a}`);
}

/**
 * Checks if the given Bencodex dictionaries have the same keys and the same
 * associated values.  In other words, this function checks if the given
 * dictionaries are encoded in the same Bencodex data.
 * @param a A dictionary to compare.
 * @param b Another dictionary to compare.
 * @returns `true` iff the given dictionaries have the same keys and the same
 *          associated values.  It returns `false` when the given dictionaries
 *          have invalid types even if they have the same keys and the same
 *          associated values, e.g., `areDictionariesEqual(1n, 1n)` returns
 *          `false`.
 */
export function areDictionariesEqual(a: Dictionary, b: Dictionary): boolean {
  if (!isDictionary(a) || !isDictionary(b)) return false;
  if (a.size !== b.size) return false;
  let otherEntries: undefined | (readonly [Key, Value])[];
  for (const [key, value] of a) {
    if (typeof key === "string") {
      const otherValue = b.get(key);
      if (typeof otherValue === "undefined") return false;
      if (!areValuesEqual(value, otherValue)) return false;
    } else {
      let otherValue = b.get(key);
      // As not every Dictionary implementation guarantees they compare binary
      // keys by their contents rather than by their references (e.g., Map),
      // we can't rely on their get() or has() methods for binary keys:
      if (typeof otherValue === "undefined") {
        if (otherEntries === undefined) otherEntries = [...b.entries()];
        for (const [bKey, bVal] of otherEntries) {
          if (areKeysEqual(key, bKey)) {
            otherValue = bVal;
            break;
          }
        }
      }
      if (typeof otherValue === "undefined") return false;
      if (!areValuesEqual(value, otherValue)) return false;
    }
  }
  return true;
}

/**
 * Checks if the given Bencodex values have the same type and the same contents.
 * In other words, this function checks if the given values are encoded in the
 * same Bencodex data.
 * @param a A Bencodex value to compare.
 * @param b Another Bencodex value to compare.
 * @returns `true` iff the given values have the same type and the same
 *          contents.  It returns `false` when the given values have invalid
 *          types even if they have the same contents, e.g., `areValuesEqual(1,
 *          1)` returns `false` (since `number`s are not a valid Bencodex type).
 */
export function areValuesEqual(a: Value, b: Value): boolean {
  if (a === null && b === null) return true;
  if (
    typeof a === "boolean" && typeof b === "boolean" ||
    typeof a === "bigint" && typeof b === "bigint"
  ) return a === b;
  if (isKey(a) && isKey(b)) return areKeysEqual(a, b);
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    return a.every((v, i) => areValuesEqual(v, b[i]));
  }
  // Note that areDictionariesEqual() checks the types of the given values
  // at runtime, and it returns false when the given values aren't dictionaries:
  return areDictionariesEqual(
    a as unknown as Dictionary,
    b as unknown as Dictionary,
  );
}
