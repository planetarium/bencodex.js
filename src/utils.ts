/**
 * Checks if the given two Uint8Arrays are equal.
 * @param a An Uint8Array.
 * @param b Another Uint8Array.
 * @returns `true` iff the given two Uint8Arrays are equal.
 */
export function areUint8ArraysEqual(a: Uint8Array, b: Uint8Array) {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

/**
 * Compares two Uint8Arrays lexicographically.
 * @param a An Uint8Array.
 * @param b Another Uint8Array.
 * @returns A negative number if `a` is less than `b`, a positive number
 *          if `a` is greater than `b`, or zero if `a` and `b` are equal.
 */
export function compareUint8Arrays(a: Uint8Array, b: Uint8Array): number {
  for (let i = 0; i < a.length && i < b.length; i++) {
    if (a[i] < b[i]) return -1;
    if (a[i] > b[i]) return 1;
  }
  if (a.length < b.length) return -1;
  if (a.length > b.length) return 1;
  return 0;
}
