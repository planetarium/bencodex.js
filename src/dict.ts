/**
 * This module provides a {@link BencodexDictionary} class which implements
 * {@link Dictionary} interface.  Although an encoder can take any object
 * implementing the {@link Dictionary} interface (e.g., a {@link Map} object),
 * a decoder always represents a Bencodex dictionary as
 * a {@link BencodexDictionary} so that it can guarantee the sensible behavior
 * according to the Bencodex specification.
 *
 * @module
 */

import {
  decode,
  encode,
} from "https://deno.land/std@0.177.0/encoding/base64.ts";
import {
  type Dictionary,
  isDictionary,
  type Key,
  type Value,
} from "./types.ts";
import { areUint8ArraysEqual } from "./utils.ts";

const SHORT_BINARY_THRESHOLD = 32;

/**
 * A {@link Dictionary} implementation that complies with the Bencodex
 * specification.  Unlike {@link Map}, this implementation allows only keys of
 * type {@link Key} and values of type {@link Value}, and in particular,
 * it actually compares {@link Uint8Array} keys by their contents instead of
 * their references.
 *
 * Note that this implementation does not guarantee the stable order of entries;
 * it can vary depending on the platform and the version of this library.
 */
export class BencodexDictionary implements Dictionary {
  readonly #stringKeys: Record<string, Value>;
  readonly #shortBinaryKeys: Record<string, Value>;
  readonly #longerBinaryKeys: [Uint8Array, Value][];
  readonly size;

  /**
   * Creates a new {@link BencodexDictionary} instance.
   * @param entries An iterable object which yields key-value pairs to be
   *                inserted into the new dictionary.  If omitted, an empty
   *                dictionary is created.
   * @throws {TypeError} When the given `entries` is not an iterable object or
   *                     when it yields a non-pair value.
   */
  constructor(entries: Iterable<[Key, Value]> = []) {
    if (typeof entries !== "object" || !(Symbol.iterator in entries)) {
      throw new TypeError(
        `Expected an iterable, but got a ${typeof entries}`,
      );
    }
    this.#stringKeys = {};
    this.#shortBinaryKeys = {};
    this.#longerBinaryKeys = [];
    this.size = 0;
    for (const pair of entries) {
      if (!Array.isArray(pair) || pair.length < 2) {
        throw new TypeError(
          "Every entry must be a pair represented as an array of length 2, " +
            (Array.isArray(pair)
              ? `but got an array of length ${pair.length}`
              : `but got a ${typeof pair}`),
        );
      }
      const [key, value] = pair;
      if (typeof key === "string") {
        if (key in this.#stringKeys) this.size--;
        this.#stringKeys[key] = value;
      } else if (!(key instanceof Uint8Array)) {
        throw new TypeError(
          `Every key must be either string or Uint8Array, but got a ${(
            typeof key
          )}`,
        );
      } else if (key.length < SHORT_BINARY_THRESHOLD) {
        const encoded = encode(key);
        if (encoded in this.#shortBinaryKeys) this.size--;
        this.#shortBinaryKeys[encoded] = value;
      } else {
        let i = 0;
        for (const [k, _] of this.#longerBinaryKeys) {
          if (areUint8ArraysEqual(k, key)) {
            break;
          }

          i++;
        }
        if (i < this.#longerBinaryKeys.length) {
          this.#longerBinaryKeys[i] = [key, value];
          continue;
        }
        this.#longerBinaryKeys.push([key, value]);
      }

      this.size++;
    }
  }

  get(key: Key): Value | undefined {
    if (typeof key === "string") return this.#stringKeys[key];
    else if (key.length < SHORT_BINARY_THRESHOLD) {
      return this.#shortBinaryKeys[encode(key)];
    }
    for (const [k, v] of this.#longerBinaryKeys) {
      if (areUint8ArraysEqual(k, key)) return v;
    }
  }

  has(key: Key): boolean {
    if (typeof key === "string") return key in this.#stringKeys;
    else if (key.length < SHORT_BINARY_THRESHOLD) {
      return encode(key) in this.#shortBinaryKeys;
    }
    for (const [k] of this.#longerBinaryKeys) {
      if (areUint8ArraysEqual(k, key)) return true;
    }
    return false;
  }

  *keys(): Iterable<Key> {
    for (const key in this.#stringKeys) {
      yield key;
    }
    for (const key in this.#shortBinaryKeys) {
      yield decode(key);
    }
    for (const [key, _] of this.#longerBinaryKeys) {
      yield key;
    }
  }

  *values(): Iterable<Value> {
    for (const value of Object.values(this.#stringKeys)) {
      yield value;
    }
    for (const value of Object.values(this.#shortBinaryKeys)) {
      yield value;
    }
    for (const [_, value] of this.#longerBinaryKeys) {
      yield value;
    }
  }

  entries(): Iterable<[Key, Value]> {
    return this;
  }

  forEach(
    callback: (value: Value, key: Key, dictionary: Dictionary) => void,
    thisArg?: unknown,
  ): void {
    for (const [key, value] of this.entries()) {
      callback.call(thisArg, value, key, this);
    }
  }

  *[Symbol.iterator](): Iterator<[Key, Value]> {
    for (const [key, value] of Object.entries(this.#stringKeys)) {
      yield [key, value];
    }
    for (const [key, value] of Object.entries(this.#shortBinaryKeys)) {
      yield [decode(key), value];
    }
    for (const pair of this.#longerBinaryKeys) {
      yield pair;
    }
  }

  [Symbol.for("Deno.customInspect")](inspect: (_: unknown) => string) {
    let s = "BencodexDictionary { ";
    let first = true;
    for (const [key, value] of this) {
      if (!first) s += ", ";
      s += `${inspect(key)}: ${inspect(value)}`;
      first = false;
    }
    s += " }";
    return s;
  }
}

/**
 * A record object that represents a Bencodex dictionary.
 *
 * Note that its values can be either {@link Value} or {@link RecordValue}
 * recursively.
 */
export interface RecordValue {
  [key: string]: Value | RecordValue;
}

/**
 * A type guard predicate that checks if the given value is
 * a {@link RecordValue}.
 * @param value The value to check.
 * @returns `true` iff the given value is a {@link RecordValue}.
 */
export function isRecordValue(
  value: Value | RecordValue,
): value is RecordValue {
  return typeof value === "object" &&
    value != null &&
    !Array.isArray(value) &&
    !(value instanceof Uint8Array) &&
    !isDictionary(value);
}

/**
 * A view of a {@link RecordValue} that implements {@link Dictionary}.
 */
export class RecordView implements Dictionary {
  #record: RecordValue;
  #size: number | undefined;
  #textEncoder: TextEncoder | undefined;
  #textDecoder: TextDecoder | undefined;

  /**
   * How the keys are encoded.  If it is `"text"`, the keys are encoded as
   * Bencodex texts.  If it is `"utf8"`, the keys are encoded as Bencodex binary
   * values with UTF-8 encoding.
   */
  readonly keyEncoding: "text" | "utf8";

  /**
   * Creates a new {@link RecordView} instance.
   * @param record The record object to view.
   * @param keyEncoding How the keys are encoded.  If it is `"text"`, the keys
   *                    are encoded as Bencodex texts.  If it is `"utf8"`, the
   *                    keys are encoded as Bencodex binary values with UTF-8
   *                    encoding.
   * @throws {TypeError} When the given record is not an object or is null.
   */
  constructor(record: RecordValue, keyEncoding: "text" | "utf8") {
    if (typeof record !== "object") {
      throw new TypeError(
        `Expected an object, but got a ${typeof record}`,
      );
    } else if (record == null) {
      throw new TypeError("Expected an object, but got null");
    }
    this.#record = record;
    this.#textEncoder = keyEncoding == "text" ? undefined : new TextEncoder();
    this.#textDecoder = keyEncoding == "text" ? undefined : new TextDecoder();
    this.keyEncoding = keyEncoding;
  }

  get size(): number {
    if (this.#size === undefined) {
      this.#size = Object.keys(this.#record).length;
    }
    return this.#size;
  }

  #getField(key: Key): string | undefined {
    if (this.#textDecoder == null) {
      return typeof key === "string" ? key : undefined;
    }

    return typeof key === "string" ? undefined : this.#textDecoder.decode(key);
  }

  get(key: Key): Value | undefined {
    const field = this.#getField(key);
    if (field == null) return undefined;
    const value = this.#record[field];
    if (isRecordValue(value)) return new RecordView(value, this.keyEncoding);
    return value;
  }

  has(key: Key): boolean {
    const field = this.#getField(key);
    return field != null && field in this.#record;
  }

  *keys(): Iterable<Key> {
    const keys = Object.keys(this.#record);
    const encoder = this.#textEncoder;
    if (encoder == null) yield* keys;
    else {
      for (const key of keys) yield encoder.encode(key);
    }
  }

  *values(): Iterable<Value> {
    for (const value of Object.values(this.#record)) {
      yield isRecordValue(value)
        ? new RecordView(value, this.keyEncoding)
        : value;
    }
  }

  entries(): Iterable<[Key, Value]> {
    return this;
  }

  forEach(
    callback: (value: Value, key: Key, dictionary: Dictionary) => void,
    thisArg?: unknown,
  ): void {
    for (const [key, value] of this) {
      callback.call(thisArg, value, key, this);
    }
  }

  *[Symbol.iterator](): Iterator<[Key, Value]> {
    const entries = Object.entries(this.#record);
    const encoder = this.#textEncoder;
    if (encoder == null) {
      for (const [key, value] of entries) {
        yield [
          key,
          isRecordValue(value)
            ? new RecordView(value, this.keyEncoding)
            : value,
        ];
      }
    } else {
      for (const [key, value] of entries) {
        yield [
          encoder.encode(key),
          isRecordValue(value)
            ? new RecordView(value, this.keyEncoding)
            : value,
        ];
      }
    }
  }

  [Symbol.for("Deno.customInspect")](inspect: (_: unknown) => string) {
    return `RecordView { ${inspect(this.#record)}`;
  }
}
