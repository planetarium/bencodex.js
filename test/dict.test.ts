import { zip } from "std/collections/zip.ts";
import {
  assert,
  assertArrayIncludes,
  assertEquals,
  assertFalse,
  assertStrictEquals,
  assertThrows,
} from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { BencodexDictionary } from "../src/dict.ts";
import { Key, Value } from "../src/types.ts";
import { areUint8ArraysEqual } from "../src/utils.ts";

Deno.test("BencodexDictionary", async (t) => {
  const dict = new BencodexDictionary([
    // string keys:
    ["foo", "bar"],
    ["baz", "qux"],
    ["foo", "dup"],
    // short binary keys:
    [new Uint8Array([0x01, 0x02, 0x03]), "bar"],
    [new Uint8Array([0x04, 0x05, 0x06]), "qux"],
    [new Uint8Array([0x01, 0x02, 0x03]), "dup"],
    // long binary keys:
    [new Uint8Array(64), "bar"],
    [new Uint8Array(64).fill(1, 0), "qux"],
    [new Uint8Array(64), "dup"],
  ]);

  await t.step("new()", () => {
    assertThrows(
      () => new BencodexDictionary(123 as unknown as Iterable<[Key, Value]>),
      TypeError,
      "got a number",
    );
    assertThrows(
      () => new BencodexDictionary(["asdf" as unknown as [Key, Value]]),
      TypeError,
      "got a string",
    );
    assertThrows(
      () => new BencodexDictionary([[] as unknown as [Key, Value]]),
      TypeError,
      "got an array of length 0",
    );
    assertThrows(
      () =>
        new BencodexDictionary(
          [[123, "invalid key"]] as unknown as Iterable<[Key, Value]>,
        ),
      TypeError,
      "string or Uint8Array",
    );
  });

  await t.step("size", () => {
    assertStrictEquals(dict.size, 6);
  });

  await t.step("get()", () => {
    assertStrictEquals(dict.get("foo"), "dup");
    assertStrictEquals(dict.get("baz"), "qux");
    assertStrictEquals(dict.get("qux"), undefined);
    assertStrictEquals(dict.get(new Uint8Array([0x01, 0x02, 0x03])), "dup");
    assertStrictEquals(dict.get(new Uint8Array([0x04, 0x05, 0x06])), "qux");
    assertStrictEquals(dict.get(new Uint8Array([0x07, 0x08, 0x09])), undefined);
    assertStrictEquals(dict.get(new Uint8Array(64)), "dup");
    assertStrictEquals(dict.get(new Uint8Array(64).fill(1, 0)), "qux");
    assertStrictEquals(dict.get(new Uint8Array(64).fill(2, 0)), undefined);
  });

  await t.step("has()", () => {
    assert(dict.has("foo"));
    assert(dict.has("baz"));
    assertFalse(dict.has("qux"));
    assert(dict.has(new Uint8Array([0x01, 0x02, 0x03])));
    assert(dict.has(new Uint8Array([0x04, 0x05, 0x06])));
    assertFalse(dict.has(new Uint8Array([0x07, 0x08, 0x09])));
    assert(dict.has(new Uint8Array(64)));
    assert(dict.has(new Uint8Array(64).fill(1, 0)));
    assertFalse(dict.has(new Uint8Array(64).fill(2, 0)));
  });

  await t.step("keys()", () => {
    const keys = [...dict.keys()];
    assertStrictEquals(keys.length, 6);
    assertArrayIncludes(keys, ["foo", "baz"]);
    const expectedBinKeys = [
      new Uint8Array([0x01, 0x02, 0x03]),
      new Uint8Array([0x04, 0x05, 0x06]),
      new Uint8Array([0x01, 0x02, 0x03]),
      new Uint8Array(64),
      new Uint8Array(64).fill(1, 0),
    ];
    for (const expectedKey of expectedBinKeys) {
      assert(
        keys.some((key) =>
          key instanceof Uint8Array && areUint8ArraysEqual(key, expectedKey)
        ),
        `expected to find ${Deno.inspect(expectedKey)}`,
      );
    }
  });

  await t.step("values()", () => {
    assertEquals([...dict.values()].sort(), [
      "dup",
      "dup",
      "dup",
      "qux",
      "qux",
      "qux",
    ]);
  });

  await t.step("entries()", () => {
    assertEquals(
      [...dict.entries()],
      zip([...dict.keys()], [...dict.values()]),
    );
  });

  await t.step("forEach()", () => {
    const entries = [...dict.entries()];
    let i = 0;
    dict.forEach((value, key, dictionary) => {
      assertStrictEquals(dictionary, dict);
      assertStrictEquals(value, entries[i][1]);
      assertEquals(key, entries[i][0]);
      i++;
    });
    assertStrictEquals(i, 6);
  });

  await t.step("[Symbol.iterator]()", () => {
    assertEquals([...dict], [...dict.entries()]);
  });

  await t.step("[Deno.customInspect]()", async (t) => {
    await assertSnapshot(t, Deno.inspect(dict));
  });
});
