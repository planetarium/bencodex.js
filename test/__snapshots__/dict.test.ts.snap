export const snapshot = {};

snapshot[`BencodexDictionary > [Deno.customInspect]() 1`] = `
'BencodexDictionary { "foo": "dup", "baz": "qux", Uint8Array(3) [ 1, 2, 3 ]: "dup", Uint8Array(3) [ 4, 5, 6 ]: "qux", Uint8Array(64) [
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
]: "dup", Uint8Array(64) [
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
]: "qux" }'
`;

snapshot[`RecordView > [Deno.customInspect]() 1`] = `'RecordView { { foo: "bar", baz: "qux", "단팥": { spam: "span", eggs: "ham" } }'`;

snapshot[`RecordView > [Deno.customInspect]() 2`] = `'RecordView { { foo: "bar", baz: "qux", "단팥": { spam: "span", eggs: "ham" } }'`;
